packages/driver/src/cypress/mocha.ts

```diff
@@ -37,6 +37,104 @@ delete (window as any).Mocha
 
 export const SKIPPED_DUE_TO_BROWSER_MESSAGE = ' (skipped due to browser)'
 
+interface CypressTest extends Mocha.Test {
+  prevAttempts: CypressTest[]
+  final?: boolean
+  forceState?: 'passed'
+}
+
+type Strategy = 'detect-flake-and-pass-on-threshold' | 'detect-flake-but-always-fail' | undefined
+
+type NormalizedRetriesConfig = {
+  strategy?: Strategy
+  maxRetries?: number
+  passesRequired?: number
+  stopIfAnyPassed?: boolean
+}
+
+// NOTE: 'calculateTestStatus' is marked as an individual function to make functionality easier to test.
+export function calculateTestStatus (test: CypressTest, config?: NormalizedRetriesConfig) {
+  // @ts-expect-error
+  const totalAttemptsAlreadyExecuted = test.currentRetry() + 1
+  let shouldAttemptsContinue: boolean = true
+  let outerTestStatus: 'passed' | 'failed' | undefined = undefined
+
+  const passedTests = _.filter(test.prevAttempts, (o) => o.state === 'passed')
```

Comment:
Could probably do this and the next line in one loop and it'd be a shade more efficient. However, guessing that prevAttempts is likely not ever that large so ultimately doesn't matter. Up to you.